ibis/tests/expr/test_case.py

```diff
@@ -177,3 +213,31 @@ def test_case_mixed_type():
     )
     result = t0[expr]
     assert result["label"].type().equals(dt.string)
+
+
+def test_err_on_nonbool(table):
+    with pytest.raises(SignatureValidationError):
+        ibis.case().when(table.a, "bar").else_("baz").end()
+
+
+@pytest.mark.xfail(reason="Literal('foo', type=bool), should error, but doesn't")
+def test_err_on_nonbool2():
+    with pytest.raises(SignatureValidationError):
+        ibis.case().when("foo", "bar").else_("baz").end()
```

Comment:
I'm going to try to fix this issue in this PR.

Reply:
This is slightly weird, because we allow conversion from `int` to `bool` during `literal` input normalization, which does match the Python behavior of e.g., `bool(1)`. This would be a breaking change, and we have a test (`test_ifelse_function_literals`) for this so I'm going to delete the test in this PR, as this behavior is intentional, there's even a comment in the test, specifically about this behavior!

Reply:
FTR this is the kind of hairy problem that we should not just start writing a bunch of tests for without exploring the issue more. It's a significant behavior change.

Reply:
I think this would be my ideal behavior:
- ibis.case().when("foo") -> error
- ibis.case().when("True") -> error
- ibis.case().when(1) -> error
- ibis.case().when(True) -> works
- ibis.case().when(ibis.literal("True", type=bool)) -> works

in general, I think we should follow the guidelines of

- If handed an ibis Expr, it must be .is_boolean()
- If handed anything else, we pass it to ibis.literal(x) (without explicitly saying `type=bool`!) and the result had better be .is_boolean(), else error.

Is this the behavior you would want? if so, we should adjust the other test to match

Reply:
I think that behavior has more cases to understand and handle than the current behavior, which is fairly easy to explain:

1. Not an ibis expression -> call the `bool` builtin on the input
2. Ibis expression -> check that it's `.type()` is boolean.

Having a third case of "call `bool` on the input only if the input type is specified is **a lot** of work to implement, including a big literal refactoring to differentiate between a type that was passed and one that wasn't with very little benefit IMO.

Also, just to make sure we're on the same page:

```
In [1]: bool("True") == bool("False")
Out[1]: True
```

Ibis tries to follow Python behavior where it makes sense.

Reply:
> 1. Not an ibis expression -> call the bool builtin on the input
> 2. Ibis expression -> check that it's .type() is boolean.

Nice, yes this is better than what I proposed. But it still does have a bit of a footgun, I don't love how `ibis.case().when("False", ...)` does opposite of what you might expect. Could we adjust the algorithm to

1. Not an ibis expression -> `assert isinstance(arg, bool)`, then do `arg = ibis.literal(arg)`
2. Ibis expression -> check that it's .type() is boolean.

It's a little more restrictive (users might need to do `case.where(bool(x))` where before they just had `case.where(x)`) and slightly breaking, but I think this might also catch some real bugs in people's code.

> Also, just to make sure we're on the same page:

Thanks, I was NOT on the same page, I mistakenly thought `bool("False") == False` and `bool("foo")` errored.

Reply:
I don't have much to back up this claim, but we haven't gotten any bug reports about this yet, but we can certainly have a discussion when they come up. I'd rather not incur a bunch of development overhead for something that might happen.

Reply:
sounds good!